tools/hangar/__snapshots__/test_corpus/valid/optional_macros.test.w_compile_tf-aws.md

```diff
@@ -0,0 +1,111 @@
+# [optional_macros.test.w](../../../../../examples/tests/valid/optional_macros.test.w) | compile | tf-aws
+
+## inflight.$Closure1-1.cjs
+```cjs
+"use strict";
+const $helpers = require("@winglang/sdk/lib/helpers");
+module.exports = function({ $expect_Util }) {
+  class $Closure1 {
+    constructor({  }) {
+      const $obj = (...args) => this.handle(...args);
+      Object.setPrototypeOf($obj, this);
+      return $obj;
+    }
+    async handle() {
+      const result = ({});
+      (await $expect_Util.equal(((result.item ?? {}))?.["id"], undefined));
+      (await $expect_Util.equal(((obj, key) => { return obj.hasOwnProperty(key); })((result.item ?? {}),"id"), false));
+      (await $expect_Util.equal((((result.items ?? []).at(0) ?? {}))?.["id"], undefined));
+      (await $expect_Util.equal(((obj, key) => { return obj.hasOwnProperty(key); })(((result.items ?? []).at(0) ?? {}),"id"), false));
+      (await $expect_Util.equal(((result.mapItem ?? new Map()))["a"], undefined));
+      (await $expect_Util.equal(("id" in ((result.mapItem ?? new Map()))), false));
+      (await $expect_Util.equal((((result.mapItems ?? []).at(0) ?? new Map()))["id"], undefined));
+      (await $expect_Util.equal(("id" in (((result.mapItems ?? []).at(0) ?? new Map()))), false));
+      (await $expect_Util.equal(result.setItem?.size, undefined));
+      (await $expect_Util.equal((await result.setItem?.has?.(6)), undefined));
+      (await $expect_Util.equal((result.setItems ?? []).at(0)?.size, undefined));
+      (await $expect_Util.equal((await (result.setItems ?? []).at(0)?.has?.(6)), undefined));
+      (await $expect_Util.equal(result.structItem?.item, undefined));
+      (await $expect_Util.equal((result.structItems ?? []).at(0)?.item, undefined));
```

Comment:
I'm not sure this is the right solution. By falling back to an empty value, like an empty array or empty object, I think it can result in the wrong runtime behavior:

```
let s: Set<num>? = nil;
let a = s?.toArray();
log(a?.length ?? -1);
```

If "a" evaluates to an empty array, the log function will print 0 instead of -1, which is a bug. "a" should evaluate to "nil" in Wing, or "undefined" in JavaScript. That way "a?.length" will evaluate to "undefined" and the log function will show "-1", like in TypeScript:

```js
class Foo {
  toArray() { return []; }
}

function makeOptionalFoo(): Foo | undefined {
  return undefined;
}

let x = makeOptionalFoo();
let y = x?.toArray();
console.log(y?.length ?? -1); // "-1"
```

I think we need the emitted code to check if the expression is null or undefined, and only run the macro code when it's not one of those values. Perhaps something like

```js
foo?.bar(args)
// ->
(($value) => $value != undefined ? <bar-macro-code> : $value)(foo, args)
```